### PR TITLE
[c10d] To make ProcessGroupNCCL to use globalStore for coordination

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117093
* #117092
* #116545
* __->__ #117075
* #117074

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -770,6 +770,9 @@ ProcessGroupNCCL::ProcessGroupNCCL(
       getCvarInt(TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC, 60 * 10 /*10 Mins*/);
   ncclTraceBufferSize_ = getCvarInt(TORCH_NCCL_TRACE_BUFFER_SIZE, 0);
   enableCollecticeHashDebug_ = (dist_debug_level_ >= DebugLevel::Detail);
+  PrefixStore* prefixStore = dynamic_cast<PrefixStore*>(store_.get());
+  globalStore_ =
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:774]: comment please! (say why we have this)

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 2]
```diff
@@ -759,6 +760,10 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // The store is used to broadcast the NCCL unique ID of rank 0.
   c10::intrusive_ptr<Store> store_;
 
+  // Reference to the store so that we can coordinate for dumping
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:763]: its not specific to coordinating dumping. its an object anyone can use for anything global.  better to say that it is a store without prefixes so keys are global across PGs.. or something.  (and clarify that store_ is prefixes)

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:774]: sure. will add more comments.

**Review by wconstab — APPROVED** [review]

**Pull request closed without merging**